modules/standard/Collection.chpl

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/*
+  A 'Collection' is a data structure, a container for elements that provide support
+  for insert, lookup, remove, and iteration operations.
+*/
+class Collection {
+  /*
+    The type of element that this Collection holds.
+  */
+  type eltType;
+
+  /*
+    Adds an element to this data structure.
+  */
+  inline proc add(elt : eltType) : bool {
+    halt("'proc add(elts : eltType ... ?nElts) : bool' is not supported...");
+  }
+
+  /*
+    Removes an arbitrary element from this data structure.
+
+    **BUG:** Compiler will segfault if the returned value is not captured at callsite.
+
+    **FIX:** Ensure that you always capture the return value...
+
+    ::
+    
+      var capturedRetval = c.remove()
+
+    **BUG:** Loop Invariant Code Motion causes undefined behavior if assigned to a
+    variable declared outside of loop
+
+    **FIX:** Use the `--no-loop-invariant-code-motion` to disable LICM.
+    Otherwise, just make sure you always capture the return value inside of a loop
+    in a variable not declared outside of loop...
+
+    ::
+
+      for i in 1 .. N {
+        var retval = c.remove();
+      }
+
+  */
+  inline proc remove() : (bool, eltType) {
+    halt("'proc remove() : (bool, eltType)' is not supported...");
+  }
+
+  /*
+    Determine whether an element exists in this collection.
+  */
+  inline proc contains(elt : eltType) : bool {
+    halt("'proc contains(elt : eltType) : bool' is not supported...");
+  }
+
+  /*
+    Clears all elements in this collection.
+  */
+  inline proc clear() {
+    while !remove()[1] do ;
+  }
+
+  /*
+    Check if this data structure is empty.
+  */
+  inline proc isEmpty() : bool {
+    return size() == 0;
+  }
+
+  /*
+    Obtain the number of elements contained in this collection.
+  */
+  inline proc size() : int {

```

Comment:
Parenthesis methods cannot be overloaded, but having a synonym for 'length' seems spot on since 'length' can just call 'size()'.